### DISCUSSION TITLE
请问为什么get_download_url在我的浏览器打开显示403或者网址迁移到新地址，而却能通过示例代码下载音频

### DISCUSSION THREAD

**Discussion by TimG233** [Nemo2011/bilibili-api#396] [Q&A] [answered]
版本：`15.5.1.b0`
我所用的`download_url`来自于示例 https://nemo2011.github.io/bilibili-api/#/examples/video
当我用
```py
async def videofunc():
    # ...
    v = video.Video(bvid=bvid, credential=credential)
    url_res = await v.get_download_url(0)

    detector = video.VideoDownloadURLDataDetecter(data=url_res)
    streams = detector.detect_best_streams()

    await download_url(streams[1].url, "audio_temp.m4s", "音频流")
```
但当我`print(streams[1].url)`并尝试去打开的时候，我会看到3钟url开头
- `https://xy183x160x126x79xy.mcdn.bilivideo.cn:8082 ...`这种可以浏览器打开，打开后直接弹出下载地址选择
- `https://cn-hbwh-cm-01-04.bilivideo.com/upgcxcode/ ...`这种显示openresty 403 forbidden
- `https://upos-sz-mirrorali.bilivideo.com/upgcxcode/`这种显示无法访问此网站
但这三种都可以下载，这个就很奇怪。希望大佬们解答下。还有就是这个视频/音频有没有可能不通过下载而直接播放出来(类似于streaming)？有没有类似的url？谢谢！

更新：后两种url似乎可以下载，但是下载的文件也打不开。是一个特别小的文件（不到1kb且不是标准音频文件，虽然以我们定的m4s结尾）

**Comment by z0z0r4** [comment]
下不了的说明有鉴权

**Reply by TimG233** [reply]
请问鉴权怎么解决，望指点

**Reply by z0z0r4** [reply, marked as answer]
参考 [视频取流说明](https://github.com/SocialSisterYi/bilibili-API-collect/blob/563dd3f7231e65564f438d43bd6c8fc6a365a235/docs/video/videostream_url.md#%E8%A7%86%E9%A2%91%E5%8F%96%E6%B5%81%E8%AF%B4%E6%98%8E) 可能对你有帮助

**Reply by TimG233** [reply]
有没有什么办法能在`get_download_url()`这步或这步之前解决这个问题，因为下载示例里`download_url`这个方法传入的HEADERS是含有referer的但这个时候也已经没用了，因为`get_download_url()`这一步给出的url就是有问题的。

**Reply by z0z0r4** [reply]
> 有没有什么办法能在`get_download_url()`这步或这步之前解决这个问题，因为下载示例里`download_url`这个方法传入的HEADERS是含有referer的但这个时候也已经没用了，因为`get_download_url()`这一步给出的url就是有问题的。

接口只提供URL，处理鉴权应该是自己的问题，与库无关

按照示例你还是没法下载吗？

**Reply by TimG233** [reply]
仍然没法下载，使用和不使用credential都是一个结果（但我发现SESSDATA， bili_jct, buvid3一天就变了，但更新后还是一样）。HEADERS是直接import的，里面也含有referer。
比如说下载这个视频`BV1M4411k7T4`，不管带不带credential都只会下载一个552bytes的文件（audio），但也打不开因为不是标准格式。而且这个视频3分多，显然就算打得开这个文件也不对。就很奇怪

**这个我晚上会继续测试，但是目前看来绝大多数可以下载，而且和credentials没关系。**

**Reply by z0z0r4** [reply]
没看懂...你 `download_url` 不是自己写的吗？

拿到下载链接了之后和 `credential` 有什么关系，自己写 UA 和 referer 写去下载，为什么和本库有关...

`credential` 会过期再正常不过了，自己重新登录或者调 refresh

**Comment by z0z0r4** [comment]
> 有没有什么办法能在`get_download_url()`这步或这步之前解决这个问题，因为下载示例里`download_url`这个方法传入的HEADERS是含有referer的但这个时候也已经没用了，因为`get_download_url()`这一步给出的url就是有问题的。

这个库只负责拿到 `https://xy183x160x126x79xy.mcdn.bilivideo.cn:8082` 这样的 URL，你要自己负责下载，鉴权信息都在上面...没理解？

`get_download_url()` 怎么可能有问题，你可能觉得是链接过期了什么的，应该只是鉴权没写对

实在不行你看看 owner 写的 [bilidown](https://github.com/Nemo2011/BiliDown) 怎么操作的

https://github.com/Nemo2011/BiliDown/blob/d2db5c8c78d3f4a13ec6917832ea7f3e6df31433/bilidown/__main__.py#L801-L849

**Reply by TimG233** [reply]
理解了，目前测起来`get_download_url()`确实没问题。谢谢。
